tfhe/src/integer/server_key/radix_parallel/ilog2.rs

```diff
@@ -0,0 +1,734 @@
+                    if (x >> i) & 1 == bit_value.opposite() as u64 {
+                        break;
+                    }
+                    count += 1;
+                }
+                count
+            }),
+            Direction::Leading => self.key.generate_lookup_table(|x| {
+                let x = x % self.key.message_modulus.0 as u64;
+
+                let mut count = 0;
+                for i in (0..self.key.message_modulus.0.ilog2()).rev() {
+                    if (x >> i) & 1 == bit_value.opposite() as u64 {
+                        break;
+                    }
+                    count += 1;
+                }
+                count
+            }),
+        };
+
+        // Assign to each block its number of leading/trailing zeros/ones
+        // in the message space
+        blocks.par_iter_mut().for_each(|block| {
+            self.key.apply_lookup_table_assign(block, &lut);
+        });
+
+        if direction == Direction::Leading {
+            // Our blocks are from lsb to msb
+            // `leading` means starting from the msb, so we reverse block
+            // for the cum sum process done later
+            blocks.reverse();
+        }
+
+        // Use hillis-steele cumulative-sum algorithm
+        // Here, each block either keeps his value (the number of leading zeros)
+        // or becomes 0 if the preceding block
+        // had a bit set to one in it (leading_zeros != num bits in message)
+        let num_bits_in_message = self.key.message_modulus.0.ilog2() as u64;
+        let sum_lut = self.key.generate_lookup_table_bivariate(
+            |block_num_leading_zeros, more_significant_block_leading_zeros| {
+                if more_significant_block_leading_zeros == num_bits_in_message {
+                    block_num_leading_zeros
+                } else {
+                    0
+                }
+            },
+        );
+
+        let sum_function =
+            |block_num_leading_zeros: &mut Ciphertext,
+             more_significant_block_leading_zeros: &Ciphertext| {
+                self.key.unchecked_apply_lookup_table_bivariate_assign(
+                    block_num_leading_zeros,
+                    more_significant_block_leading_zeros,
+                    &sum_lut,
+                );
+            };
+        self.compute_prefix_sum_hillis_steele(blocks, sum_function)
+    }
+
+    /// Counts how many consecutive bits there are
+    fn count_consecutive_bits<T>(
+        &self,
+        ct: &T,
+        direction: Direction,
+        bit_value: BitValue,
+    ) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        if ct.blocks().is_empty() {
+            return self.create_trivial_zero_radix(ct.blocks().len());
+        }
+
+        let num_bits_in_message = self.key.message_modulus.0.ilog2();
+        let original_num_blocks = ct.blocks().len();
+
+        let num_bits_in_ciphertext = num_bits_in_message
+            .checked_mul(original_num_blocks as u32)
```

Comment:
are we expecting to count upwards of 4 billions 1s and 0s ? 👀

Reply:
No, that's why I put checked_mul

Reply:
misread checked_mul my bad